encoding/dot/decode.go

```diff
@@ -0,0 +1,232 @@
+package dot
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/gonum/graph"
+	"github.com/graphism/dot/ast"
+	"github.com/pkg/errors"
+)
+
+// Builder is a graph that can have user-defined nodes and edges added.
+type Builder interface {
+	graph.Graph
+	graph.Builder
+	// NewNode adds a new node with a unique node ID to the graph.
+	NewNode() graph.Node
+	// NewEdge adds a new edge from the source to the destination node to the
+	// graph, or returns the existing edge if already present.
+	NewEdge(from, to graph.Node) graph.Edge
+}
+
+// UnmarshalerAttr is the interface implemented by objects that can unmarshal a
+// DOT attribute description of themselves.
+type UnmarshalerAttr interface {
+	// UnmarshalDOTAttr decodes a single DOT attribute.
+	UnmarshalDOTAttr(attr Attribute) error
+}
+
+// Copy copies the nodes and edges from the Graphviz AST source graph to the
+// destination graph. Edge direction is maintained if present.
+func Copy(dst Builder, src *ast.Graph) error {
+	gen := &generator{
+		directed: src.Directed,
+		ids:      make(map[string]graph.Node),
+	}
+	for _, stmt := range src.Stmts {
+		gen.addStmt(dst, stmt)
+	}
+	if len(gen.errs) > 0 {
+		return gen.errs
+	}
+	return nil
+}
+
+// A generator keeps track of the information required for generating a gonum
+// graph from a dot AST graph.
+type generator struct {
+	// Directed graph.
+	directed bool
+	// Map from dot AST node ID to gonum node.
+	ids map[string]graph.Node
+	// Nodes processed within the context of a subgraph, that is to be used as a
+	// vertex of an edge.
+	subNodes []graph.Node
+	// Stack of start indices into the subgraph node slice. The top element
+	// corresponds to the start index of the active (or inner-most) subgraph.
+	subStart []int
+	// List of errors encountered during decoding.
+	errs errlist
+}
+
+// node returns the gonum node corresponding to the given dot AST node ID,
+// generating a new such node if none exist.
+func (gen *generator) node(dst Builder, id string) graph.Node {
+	if n, ok := gen.ids[id]; ok {
+		return n
+	}
+	n := dst.NewNode()
+	gen.ids[id] = n
+	// Check if within the context of a subgraph, that is to be used as a vertex
+	// of an edge.
+	if gen.isInSubgraph() {
+		// Append node processed within the context of a subgraph, that is to be
+		// used as a vertex of an edge
+		gen.appendSubgraphNode(n)
+	}
+	return n
+}
+
+// addStmt adds the given statement to the graph.
+func (gen *generator) addStmt(dst Builder, stmt ast.Stmt) {
+	switch stmt := stmt.(type) {
+	case *ast.NodeStmt:
+		n := gen.node(dst, stmt.Node.ID)
+		if n, ok := n.(UnmarshalerAttr); ok {
+			for _, attr := range stmt.Attrs {
+				a := Attribute{
+					Key:   attr.Key,
+					Value: attr.Val,
+				}
+				if err := n.UnmarshalDOTAttr(a); err != nil {
+					gen.errs = append(gen.errs, errors.Wrapf(err, "unable to unmarshal node DOT attribute (%v)", a))
```

Comment:
Here I'd be inclined to have a specific error type that holds the kinds of information you want and has methods to retain the issues raised during the parse - something like os.PathError but with methods (possibly even a Cause method). I'm not convinced that retaining all errors is what we want though. You can just panic here, and recover the panic and return the error.